### ISSUE TITLE
Stop SEED VM, restart it and run nodetool repair (with background load) -> all c-s loaders fail upon triggering repair

### ISSUE THREAD

**Issue by tomer-sandler** [scylladb/scylla#2760]
*Installation details*
Scylla version (or git commit hash): 2.0 RC3
Cluster size: 15 nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): AWS AMI

*Hardware details (for performance issues)*
Platform (physical/VM/cloud instance type/docker): VM i3.4XL
Disks: 2 x NVMe per VM, each 1900GB

*Scenario*
1. Using c-s default schema (RF=3 CL=one) populating 1TB data

Schema:
```
CREATE KEYSPACE keyspace1 WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '3'}  AND durable_writes = true;

CREATE TABLE keyspace1.standard1 (
    key blob PRIMARY KEY,
    "C0" blob,
    "C1" blob,
    "C2" blob,
    "C3" blob,
    "C4" blob
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.1
    AND speculative_retry = '2.00ms';
```
```
#nohup cassandra-stress write no-warmup n=340000000 cl=one -mode native cql3 -node [IPs] -rate threads=200 -log file=populate_1TB_1-4.log -pop seq=1..340000000 &
#nohup cassandra-stress write no-warmup n=340000000 cl=one -mode native cql3 -node [IPs] -rate threads=200 -log file=populate_1TB_2-4.log -pop seq=340000001..680000000 &
#nohup cassandra-stress write no-warmup n=340000000 cl=one -mode native cql3 -node [IPs] -rate threads=200 -log file=populate_1TB_3-4.log -pop seq=680000001..1020000000 &
#nohup cassandra-stress write no-warmup n=340000000 cl=one -mode native cql3 -node [IPs] -rate threads=200 -log file=populate_1TB_4-4.log -pop seq=1020000001..1360000000 &
```

2. Waiting for compactions to end -> Ended.

3. Decommission one node without any background traffic -> completed within 8 min -> Cluster ring is now 14 nodes

4. Run 50% mixed load, CL=Q, from 8 c-s instanceds (4 VMs, 2 instances on each VM) to get steady 200K SLA

```
#nohup cassandra-stress mixed ratio\(write=1,read=1\) no-warmup duration=30m cl=quorum -pop dist=gaussian\(1..1360000000,680000000,227000000\) -mode native cql3 -node [IPs] -log file=50_mixed_nodefail_repair -rate threads=150 limit=25000/s &
```

5. After ~5 min, stop the SEED node VM (172.16.0.45) -> SLA kept, even climbed a bit to ~213K

6. After VM down for ~5 min -> restart the VM. Once it's up run ```nodetool repair -pr keyspace1```

7. Upon trigerring repair, all c-s loaders stopped with the following error/s:
```
java.io.IOException: Operation x0 on key(s) [4e31394e4e35384f3431]: Data returned was not validated

        at org.apache.cassandra.stress.Operation.error(Operation.java:135)
        at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:113)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:98)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:106)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:258)
        at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:321)
java.io.IOException: Operation x1 on key(s) [3437374e4e37324f3230]: Error executing: (ReadTimeoutException): Cassandra timeout during read query at consistency QUORUM (3 responses were required but only 2 replica responded)

        at org.apache.cassandra.stress.Operation.error(Operation.java:135)
        at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:113)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:98)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:106)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:258)
        at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:321)
java.io.IOException: Operation x1 on key(s) [333934304c4c324f5030]: Error executing: (ReadTimeoutException): Cassandra timeout during read query at consistency QUORUM (3 responses were required but only 2 replica responded)

        at org.apache.cassandra.stress.Operation.error(Operation.java:135)
        at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:113)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:98)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:106)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:258)
        at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:321)
FAILURE
```
OR
```
java.io.IOException: Operation x0 on key(s) [37373950373339333730]: Data returned was not validated

        at org.apache.cassandra.stress.Operation.error(Operation.java:135)
        at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:113)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:98)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:106)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:258)
        at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:321)
```

![startback_seed_node_run_repair_all_loaders_fail](https://user-images.githubusercontent.com/25585995/30018900-5a1a0874-9167-11e7-84cc-b86d3bc36f23.PNG)

Logs attached.
[172.16.0.4_scylla-logs.txt.gz](https://github.com/scylladb/scylla/files/1274333/172.16.0.4_scylla-logs.txt.gz)
[172.16.0.45_scylla-logs.txt.gz](https://github.com/scylladb/scylla/files/1274331/172.16.0.45_scylla-logs.txt.gz)
[172.16.0.92_scylla-logs.txt.gz](https://github.com/scylladb/scylla/files/1274332/172.16.0.92_scylla-logs.txt.gz)

**Comment by tomer-sandler** [comment]
No Core dumps in any of the 14 nodes.

**Comment by tomer-sandler** [comment]
Found the following print in 2 of the 14 nodes:
```
Sep 04 08:18:17 ip-172-16-0-4 scylla[12679]:  [shard 0] rpc - client 172.16.0.45: fail to connect: Connection refused
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 0] gossip - InetAddress 172.16.0.45 is now UP, status = NORMAL
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 0] storage_service - Node 172.16.0.45 state jump to normal
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 0] compaction - Compacting [/var/lib/scylla/data/system/peers-37f71aca7dc2383ba70672528af04d4f/system-peers-ka-98-Data.db:level=0, /var/lib/scylla/data/system/peers
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 1] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 11] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 3] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 12] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 12] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 8] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 10] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 5] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 5] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 3] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 9] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 7] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 1] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 0] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 12] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 5] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 4] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 11] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 6] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 11] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
Sep 04 08:18:18 ip-172-16-0-4 scylla[12679]:  [shard 7] storage_proxy - Exception when communicating with 172.16.0.45: Can't find a column family with UUID 0b155ea0-9076-11e7-aa85-000000000003
```

**Comment by tgrabiec** [comment]
@tomer-sandler As for the `Can't find a column family with UUID ..` message, did you create the schema before starting concurrent loaders, or not? If loaders try to create the table concurrently, you may have hit https://github.com/scylladb/scylla/issues/420. Solution is to create the schema before starting the test.

**Comment by tomer-sandler** [comment]
Yes @tgrabiec , I'm aware of the Schema to be created prior to running the load.
The Schema was created before the initial DB population... in today's test I was using the system after population (1TB), running on the node decommission and later the 50% mix load using the same Schema, while performing the node stop -> start and repair.

So this does not explain these errors.

**Comment by tgrabiec** [comment]
@tomer-sandler Looks like the seed (`ip-172-16-0-45`) didn't know about the table when the errors were logged. The table was created on it later, probably after automatic schema sync:

```
Sep 04 08:18:18 ip-172-16-0-45.ec2.internal scylla[5039]:  [shard 0] schema_tables - Creating keyspace1.standard1 id=0b155ea0-9076-11e7-aa85-000000000003 version=b4d95a1e-fa61-37c6-973f-9849eaac77d6
```

This is weird, because if you stopped it after the schema was created, the node should know about the table. But the log you attached doesn't go past the restart (starts at `08:18:17`), so I can't tell. Are you sure you restarted it and started a brand new node? Do you have its logs from before the restart?

**Comment by tzach** [comment]
Restarting a VM on EC2 will give it a new internal IP, one that does not appear on peer table.
If this is the case, we might need to restart the service with replace_address
http://docs.scylladb.com/procedures/replace_dead_node/

**Comment by gleb-cloudius** [comment]
> Restarting a VM on EC2 will give it a new internal IP.
I do not believe this is the case. External yes, but not internal.

--
			Gleb.

**Comment by tzach** [comment]
> I do not believe this is the case. External yes, but not internal.

Yep, you are right.

**Comment by tomer-sandler** [comment]
I can confirm what @gleb-cloudius wrote. Stop start on AWS does NOT change the internal IP, but the external IP changes.
@tgrabiec I took all the logs I had on the seed, perhaps the stop-start VM cleans the journalctl...

**Comment by tzach** [comment]
> perhaps the stop-start VM cleans the journalctl...

Should not. what command are you using to extract the log?

**Comment by tomer-sandler** [comment]
@tzach I was using the node_health_check to collect the logs.
I just reproduced it again now on a NON-seed node (172.16.0.203)
15 node cluster -> stopped 172.16.0.203 for ~5 min -> start it again, once it's in status ```UN```, start repair on it ```nodetool repair -pr keyspace1``` -> same error on all loaders.

@tgrabiec - can you have a look at it now?

**Comment by tomer-sandler** [comment]
It appears that in our AMI journalctl is configured to save logs only from last boot. This will make debugging hard on issues envolving a node down-up (as in my case here).
I opened a ticket on it: https://github.com/scylladb/scylla/issues/2763

**Comment by tgrabiec** [comment]
@tomer-sandler It looks like you're using ephemeral drives on that instance, so when you start the node back, it has no data.

**Comment by tgrabiec** [comment]
It still doesn't explain why you get errors at CL=QUORUM.

**Comment by gleb-cloudius** [comment]
On Tue, Sep 05, 2017 at 10:49:48AM +0000, Tomasz Grabiec wrote:
> It still doesn't explain why you get errors at CL=QUORUM.
> 
He does not start to get them until he runs repair. May be the repair
is actually corrupting data somehow when the cluster is in the undefined
state after all system tables on one node where removed. This will cause
read repair to kick in and get errors we see given that restarted node
fail reads because of schema mismatch.

--
			Gleb.

**Comment by gleb-cloudius** [comment]
Also "java.io.IOException: Operation x0 on key(s) [4e31394e4e35384f3431]: Data returned was not validated" errors do point to data corruption.

**Comment by tomer-sandler** [comment]
I should mention that in the 2 times I hit this issue, it was after streaming was done in the cluster, either to add a new node (with auto_bootstrap = true), or remove a node by ```nodetool decommission``` command.
in other cases without add node / remove node even when I stop-> start the VM and start the repair, there were no errors on the c-s loaders and repair finished successfully. If all data was lost, or data corruption occurred, how is that possible?

**Comment by tgrabiec** [comment]
2017-09-05 12:58 GMT+02:00 Gleb Natapov <notifications@github.com>:

> On Tue, Sep 05, 2017 at 10:49:48AM +0000, Tomasz Grabiec wrote:
> > It still doesn't explain why you get errors at CL=QUORUM.
> >
> He does not start to get them until he runs repair. May be the repair
> is actually corrupting data somehow when the cluster is in the undefined
> state after all system tables on one node where removed. This will cause
> read repair to kick in and get errors we see given that restarted node
> fail reads because of schema mismatch.
>

I wonder how could corruption happen though. Repair can normally only add
writes, not subtract them, so it shouldn't make things worse.

**Comment by slivne** [comment]
@tzach @gleb-cloudius 

 * if the node is using ec2 classic and the node is stopped it will loose both internal and external ips
 * if the node is in a vpc and the node is stopped it will loose only the external ip

So when the node is started backup it starts with the same internal ip and connects to the previous running cluster - yet its connecting in a wrong state - the node is connecting in a bootstrap mode trying to join the cluster and the cluster "thinks" this node is already part of it. 

the issue is that this should not be done - the node that comes up generates a new set of tokens and a new uuid - as it has no idea it was already part of the cluster in the past and is added with them into the cluster.

this causes a mess in the cluster - as its not a newly added node and we are changing ownership of token ranges.

there is a method to add a node back into the cluster - using the replace-address in that case the node added back will get the tokens (and the data) from the other nodes. 

we maybe could detect this specific case - a bootstrapping node being added with the same ip and not via replace-address procedure.

in my view its a low priority item.

**Comment by tzach** [comment]
@slivne sounds like a good explanation for the issue. Lets first validate this was indeed the case.

**Comment by tzach** [comment]
@nirmaayan We need to add the above info to Scylla procedure doc.
